wfe2/verify.go

```diff
@@ -573,7 +567,7 @@ func (wfe *WebFrontEndImpl) validJWSForAccount(
 	ctx context.Context,
 	logEvent *web.RequestEvent) ([]byte, *jose.JSONWebSignature, *core.Registration, *probs.ProblemDetails) {
 	// Lookup the account and JWK for the key ID that authenticated the JWS
-	pubKey, account, prob := wfe.lookupJWK(jws, ctx, request, logEvent)
+	pubKey, account, prob := wfe.lookupJWK(jws.Signatures[0].Header, ctx, request, logEvent)
```

Comment:
So the issue here is that we've pushed the assumption up one level. This function doesn't know where its jws came from either, so it doesn't know for sure that it has a Signatures slice of length exactly 1.

The same is true for validSelfAuthenticatedJWS and validKeyRollover, below.

So either the "just take a header" solution you've gone with here needs to be pushed even further up the call stack, or we need to take a different approach. I like the header approach in general, but I fear that it breaks down here because you need the actual jws for the call to `validJWSForKey` on line 576, so I don't have a strong opinion on which is better.